File: DSA-C++/6-array/19-print-max-subarray-sum.cpp
#include <bits/stdc++.h>

using namespace std;
// Given an integer array nums, find the subarray with the largest sum and print all elements present in subarray.


// brute force - O(n²), O(1)
void print_max_subarray_sum_1(vector <int> &nums) {
    int n = nums.size(), start, end, max_sum;
    start = end = 0;
    max_sum = INT_MIN;

    if (nums.empty()) return;

    for (int i=0; i<n; ++i) {
        int sum = 0;

        for (int j=i; j<n; ++j) {
            sum += nums[j];

            if (sum > max_sum) {
                max_sum = sum;
                start = i;
                end = j;
            }
        }
    }

    for (int i=start; i<=end; ++i) {
        cout << nums[i] << " ";
    }
    
    cout << endl;
}


// kadane - O(n), O(1)
void print_max_subarray_sum_2(vector <int> &nums) {
    int n = nums.size(), sum, max_sum, start, temp_start, end;
    start = temp_start = end = sum = 0;
    max_sum = INT_MIN;

    if (nums.empty()) return;

    for (int i=0; i<n; ++i) {
        sum += nums[i];

        if (sum > max_sum) {
            max_sum = sum;
            start = temp_start;
            end = i;
        }

        if (sum < 0) {
            sum = 0;
            temp_start = i+1;
        }
    }

    for (int i=start; i<=end; ++i) {
        cout << nums[i] << " ";
    }
 
    cout << endl;
}

int main() {
    vector <int> nums = {2, 3, 5, -2, 7, -4};
    nums = {-2, 1, -3, 4, -1, 2, 1, -5, 4};
    // nums = {-3, -2, -5};
    nums = {-2, 0, -1};
    // nums = {};
    
    print_max_subarray_sum_1(nums);

    print_max_subarray_sum_2(nums);

    return 0;
}

File: DSA-Python/5-array/19-print-max-subarray-sum.py
# Given an integer array nums, find the subarray with the largest sum, and print the subarray



# brute force - O(n²), O(1)
def max_subarray_sum_1(nums:list[int]) -> None:
    n , max_sum = len(nums), float('-inf')
    start, end = 0, 0
    
    for i in range(n):
        sum = 0
        
        for j in range(i, n):
            sum += nums[j]
            
            if sum > max_sum:
                max_sum = sum
                start = i
                end = j+1
                
    for i in range(start, end):
        print(nums[i], end=' ')
    print()



# kadane - O(n), O(1)
def max_subarray_sum_2(nums:list[int]) -> None:
    n, max_sum, sum = len(nums), float('-inf'), 0
    start, temp_start, end = 0, 0, 0
    
    
    for i in range(n):
        sum += nums[i]
        
        if sum > max_sum:
            max_sum = sum
            start = temp_start
            end = i+1
            
        if sum < 0:
            sum = 0
            temp_start = i+1
            
    for i in range(start, end):
        print(nums[i], end=' ')
    print()



if __name__ == "__main__":
    loop = int(input())
    
    for _ in range(loop):
        nums = list(map(int, input().split()))
        
        # max_subarray_sum_1(nums)
        
        max_subarray_sum_2(nums)
        
        
'''
5
-8 -3 -6 -2 -5 -4
0 -3 0 -2 0
-2 1 -3 4 -1 2 1 -5 4 
1
5 4 -1 7 8
'''